File: examples/Delphi/TXTVIEW1.pas
{$J+} {Writable constants}

unit Txtview1;

interface

uses
  SysUtils, WinTypes, WinProcs, Messages, Classes, Graphics, Controls, Forms,
  Dialogs, ExtCtrls, OvcBase, OvcViewr, OvcData, Menus, Printers, ShellAPI,
  ComCtrls;

type
  TViewForm = class(TForm)
    MainMenu1: TMainMenu;
    File1: TMenuItem;
    Exit1: TMenuItem;
    N1: TMenuItem;
    PrintSetup1: TMenuItem;
    Print1: TMenuItem;
    N2: TMenuItem;
    Open1: TMenuItem;
    Clipboard1: TMenuItem;
    SelectAll1: TMenuItem;
    Copy2: TMenuItem;
    Search1: TMenuItem;
    GotoLine1: TMenuItem;
    N3: TMenuItem;
    FindNext1: TMenuItem;
    Find1: TMenuItem;
    Options1: TMenuItem;
    Font1: TMenuItem;
    N4: TMenuItem;
    TabSize1: TMenuItem;
    N5: TMenuItem;
    WordStarCommands1: TMenuItem;
    TabExpansion1: TMenuItem;
    Help1: TMenuItem;
    About1: TMenuItem;
    DefaultController: TOvcController;
    FontDialog1: TFontDialog;
    OpenDialog1: TOpenDialog;
    PrinterSetupDialog1: TPrinterSetupDialog;
    PrintDialog1: TPrintDialog;
    Viewer1: TOvcTextFileViewer;
    Copy1: TMenuItem;
    StatusBar1: TStatusBar;
    ReplaceDialog1: TReplaceDialog;
    procedure Font1Click(Sender: TObject);
    procedure Open1Click(Sender: TObject);
    procedure Exit1Click(Sender: TObject);
    procedure PrintSetup1Click(Sender: TObject);
    procedure Print1Click(Sender: TObject);
    procedure SelectAll1Click(Sender: TObject);
    procedure Clipboard1Click(Sender: TObject);
    procedure File1Click(Sender: TObject);
    procedure CopyClick(Sender: TObject);
    procedure Search1Click(Sender: TObject);
    procedure GotoLine1Click(Sender: TObject);
    procedure Viewer1ShowStatus(Sender: TObject; Line: Longint;
      Col: Integer);
    procedure About1Click(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure TabExpansion1Click(Sender: TObject);
    procedure TabSize1Click(Sender: TObject);
    procedure Find1Click(Sender: TObject);
    procedure FindNext1Click(Sender: TObject);
    procedure WordStarCommands1Click(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
  private
    { Private declarations }
    procedure AppMessage(var Msg: TMsg; var Handled: Boolean);
  public
    { Public declarations }
  end;

var
  ViewForm: TViewForm;

implementation

uses TxtView2, TxtView3;

{$R *.DFM}

procedure TViewForm.AppMessage(var Msg: TMsg; var Handled: Boolean);
var
  hDrop : THandle;
  DroppedName : array[0..80] of Char;
begin
  if Msg.Message = WM_DROPFILES then begin
    Handled := True;
    hDrop := Msg.wParam;
    DragQueryFile(hDrop, 0, DroppedName, SizeOf(DroppedName));
    Viewer1.FileName := StrPas(DroppedName);
    Viewer1.IsOpen := True;
    DragFinish(hDrop);
    if IsIconic(Application.Handle) then
      ShowWindow(Application.Handle, sw_ShowNormal)
    else
      BringWindowToTop(Handle);
  end;
end;

procedure TViewForm.Font1Click(Sender: TObject);
begin
  FontDialog1.Font.Assign(Viewer1.FixedFont.Font);                     {!!.10}
  if FontDialog1.Execute then
    if FontDialog1.Font <> nil then
      Viewer1.FixedFont.Assign(FontDialog1.Font);
end;

procedure TViewForm.Open1Click(Sender: TObject);
begin
  if OpenDialog1.Execute then begin
    with Viewer1 do begin
      FileName := OpenDialog1.Filename;
      IsOpen := True;
    end;
    ViewForm.Caption := 'Text View - ' + OpenDialog1.Filename;
    Search1Click(Self);
    Clipboard1Click(Self);
  end;
end;

procedure TViewForm.Exit1Click(Sender: TObject);
begin
  Close;
end;

procedure TViewForm.PrintSetup1Click(Sender: TObject);
begin
  PrinterSetupDialog1.Execute;
end;

procedure TViewForm.Print1Click(Sender: TObject);
var
  Line: LongInt;
  Buf: array[0..1023] of Char;
  SelRange: TOvcViewerRange;
  PrintText: System.Text;
begin
  if Viewer1.GetSelection(SelRange.Start.Line, SelRange.Start.Col,
             SelRange.Stop.Line, SelRange.Stop.Col) then
    PrintDialog1.Options := PrintDialog1.Options + [poSelection]
  else PrintDialog1.Options := PrintDialog1.Options - [poSelection];
  if PrintDialog1.Execute then begin
    AssignPrn(PrintText);
    try
      Rewrite(PrintText);
      Printer.Canvas.Font := Viewer1.FixedFont.Font;
      case PrintDialog1.PrintRange of
        prAllPages :
          for Line := 0 to Pred(Viewer1.LineCount) do begin
            Viewer1.GetPrintableLine(Line, Buf, SizeOf(Buf));
            Writeln(PrintText, Buf);
          end;
        prSelection :
          for Line := SelRange.Start.Line to SelRange.Stop.Line do begin
            Viewer1.GetPrintableLine(Line, Buf, SizeOf(Buf));
            if Line = SelRange.Start.Line then
              FillChar(Buf, SelRange.Start.Col, #32);
            if Line = SelRange.Stop.Line then
              Buf[SelRange.Stop.Col] := #0;
            Writeln(PrintText, Buf);
          end;
      end;
    finally
      System.Close(PrintText);
    end;
  end;
end;

procedure TViewForm.SelectAll1Click(Sender: TObject);
begin
  Viewer1.SelectAll(False);
end;

procedure TViewForm.Clipboard1Click(Sender: TObject);
var
  SelRange: TOvcViewerRange;
begin
  SelectAll1.Enabled := Viewer1.LineCount > 0;
  Copy1.Enabled := Viewer1.GetSelection(SelRange.Start.Line, SelRange.Start.Col,
          SelRange.Stop.Line, SelRange.Stop.Col);
  Copy2.Enabled := Viewer1.GetSelection(SelRange.Start.Line, SelRange.Start.Col,
          SelRange.Stop.Line, SelRange.Stop.Col);
end;

procedure TViewForm.File1Click(Sender: TObject);
begin
  Print1.Enabled := Viewer1.LineCount > 0;
end;

procedure TViewForm.CopyClick(Sender: TObject);
begin
  Viewer1.CopyToClipboard;
end;

procedure TViewForm.Search1Click(Sender: TObject);
begin
  Find1.Enabled := Viewer1.LineCount > 0;
  FindNext1.Enabled := (Viewer1.LineCount > 0) and (FindDlg.Edit1.Text <> '');
  GoToLine1.Enabled := Viewer1.LineCount > 0;
end;

procedure TViewForm.GotoLine1Click(Sender: TObject);
var
  S : string;
  GoPos : TOvcTextPos;
begin
  S := '';
  if InputQuery('Go to Line Number', 'Enter Line Number (1 to ' +
    IntToStr(Viewer1.LineCount) + ')', S) then begin
    try
      GoPos.Line := Pred(StrToInt(S));
    except
      Exit;
    end;
    GoPos.Col := 0;
    Viewer1.CaretActualPos := GoPos;
  end;
end;

procedure TViewForm.Viewer1ShowStatus(Sender: TObject; Line: Longint;
  Col: Integer);
begin
  StatusBar1.Panels[1].Text := IntToStr(Succ(Viewer1.CaretEffectivePos.Line)) +
    ':' + IntToStr(Succ(Viewer1.CaretEffectivePos.Col));
  StatusBar1.Panels[2].Text := 'Total: ' + IntToStr(Viewer1.LineCount);
  StatusBar1.Panels[3].Text := 'Top: ' + IntToStr(Succ(Viewer1.TopLine));
  StatusBar1.Panels[4].Text := 'Bytes: ' + IntToStr(Viewer1.FileSize);
  Clipboard1Click(Self);
end;

procedure TViewForm.About1Click(Sender: TObject);
begin
  ViewAbout.ShowModal;
end;

procedure TViewForm.FormCreate(Sender: TObject);
begin
  TabExpansion1.Checked := Viewer1.ExpandTabs;
  DragAcceptFiles(Application.Handle, True);
  DragAcceptFiles(Handle, True);
  Application.OnMessage := AppMessage;
end;

procedure TViewForm.TabExpansion1Click(Sender: TObject);
begin
  TabExpansion1.Checked := not(TabExpansion1.Checked);
  Viewer1.ExpandTabs := TabExpansion1.Checked;
end;

procedure TViewForm.TabSize1Click(Sender: TObject);
var
  S : string;
begin
  S := IntToStr(Viewer1.TabSize);
  if InputQuery('Tab Size', 'Enter Tab Size', S) then try
    Viewer1.TabSize := StrToInt(S);
  except
    Exit;
  end;
end;

procedure TViewForm.Find1Click(Sender: TObject);
var
  OptionSet : TSearchOptionSet;
begin
  FindDlg.ShowModal;
  if FindDlg.ModalResult=mrOK then begin
    OptionSet := [];
    if FindDlg.RadioButton2.Checked then
      OptionSet := OptionSet + [soBackward];
    if FindDlg.Checkbox1.Checked then
      OptionSet := OptionSet + [soMatchCase];
    if FindDlg.Checkbox2.Checked then
      OptionSet := OptionSet + [soGlobal];
    if not Viewer1.Search(FindDlg.Edit1.Text, OptionSet) then
      MessageDlg('Search string ''' + FindDlg.Edit1.Text +
        ''' not found', mtInformation, [mbOk], 0);
    Search1Click(Self);
  end;
end;

procedure TViewForm.FindNext1Click(Sender: TObject);
var
  OptionSet : TSearchOptionSet;
begin
  OptionSet := [];
  if FindDlg.RadioButton2.Checked then
    OptionSet := OptionSet + [soBackward];
  if FindDlg.Checkbox1.Checked then
    OptionSet := OptionSet + [soMatchCase];
  if not Viewer1.Search(FindDlg.Edit1.Text, OptionSet) then
    MessageDlg('Search string ''' + FindDlg.Edit1.Text +
      ''' not found', mtInformation, [mbOk], 0);
end;

procedure TViewForm.WordStarCommands1Click(Sender: TObject);
begin
  with WordStarCommands1 do begin
    Checked := not(Checked);
    if Checked then
      DefaultController.EntryCommands.Table[0].IsActive := True
    else
      DefaultController.EntryCommands.Table[0].IsActive := False;
  end;
end;

procedure TViewForm.FormDestroy(Sender: TObject);
begin
  DragAcceptFiles(Application.Handle, False);
  DragAcceptFiles(Handle, False);
end;

end.

File: examples/CBuildr5/TxtView1.cpp
//---------------------------------------------------------------------------
#include <vcl\vcl.h>
#pragma hdrstop

#include <shellapi.h>

#include "TxtView1.h"
#include "TxtView2.h"
//---------------------------------------------------------------------------
#pragma link "OvcViewr"
#pragma link "OvcBase"
#pragma resource "*.dfm"
TViewForm *ViewForm;
//---------------------------------------------------------------------------
__fastcall TViewForm::TViewForm(TComponent* Owner)
  : TForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TViewForm::AppMessage(TMsg& Msg, bool& Handled)
{
  HDROP hDrop;
  char DroppedName[MAX_PATH];
  if (Msg.message == WM_DROPFILES) {
    Handled = true;
    hDrop = (HDROP)Msg.wParam;
    DragQueryFile(hDrop, 0, DroppedName, sizeof(DroppedName));
    Viewer1->FileName = DroppedName;
    Viewer1->IsOpen = true;
    DragFinish(hDrop);
    if (IsIconic(Application->Handle))
      ShowWindow(Application->Handle, SW_SHOWNORMAL);
    else
      BringWindowToTop(Handle);
  }
}

void __fastcall TViewForm::Font1Click(TObject *Sender)
{
  FontDialog1->Font->Assign(Viewer1->FixedFont->Font);
  if (FontDialog1->Execute())
    Viewer1->FixedFont->Assign(FontDialog1->Font);
}
//---------------------------------------------------------------------------
void __fastcall TViewForm::Open1Click(TObject *Sender)
{
  if (OpenDialog1->Execute()) {
    Viewer1->FileName = OpenDialog1->FileName;
    Viewer1->IsOpen = true;
    Caption = "Text Edit - " + OpenDialog1->FileName;
    Search1Click(this);
    Clipboard1Click(this);
  }
}
//---------------------------------------------------------------------------
void __fastcall TViewForm::Exit1Click(TObject *Sender)
{
  Close();  
}
//---------------------------------------------------------------------------
void __fastcall TViewForm::PrintSetup1Click(TObject *Sender)
{
  PrinterSetupDialog1->Execute();
}
//---------------------------------------------------------------------------
void __fastcall TViewForm::SelectAll1Click(TObject *Sender)
{
  Viewer1->SelectAll(false);
}
//---------------------------------------------------------------------------
void __fastcall TViewForm::Clipboard1Click(TObject *Sender)
{
  TOvcViewerRange SelRange;
  SelectAll1->Enabled = Viewer1->LineCount > 0;
  Copy1->Enabled = Viewer1->GetSelection(SelRange.Start.Line, SelRange.Start.Col,
          SelRange.Stop.Line, SelRange.Stop.Col);
  Copy2->Enabled = Viewer1->GetSelection(SelRange.Start.Line, SelRange.Start.Col,
          SelRange.Stop.Line, SelRange.Stop.Col);
}
//---------------------------------------------------------------------------
void __fastcall TViewForm::File1Click(TObject *Sender)
{
  Print1->Enabled = Viewer1->LineCount > 0;
}
//---------------------------------------------------------------------------
void __fastcall TViewForm::Copy1Click(TObject *Sender)
{
  Viewer1->CopyToClipboard();
}
//---------------------------------------------------------------------------
void __fastcall TViewForm::Search1Click(TObject *Sender)
{
  Find1->Enabled = Viewer1->LineCount > 0;
  FindNext1->Enabled =
    (Viewer1->LineCount > 0) && (FindDialog1->FindText != "");
  GotoLine1->Enabled = Viewer1->LineCount > 0;
}
//---------------------------------------------------------------------------
void __fastcall TViewForm::GotoLine1Click(TObject *Sender)
{
  String S = "1";
  if (InputQuery("Go to Line Number", "Enter Line Number (1 to " +
    String((int)Viewer1->LineCount) + String(")"), S))
    try {
      Viewer1->SetCaretPosition(S.ToInt(), 1);
    }
    catch (...) {
      return;
    }
}
//---------------------------------------------------------------------------
void __fastcall TViewForm::Viewer1ShowStatus(TObject *Sender, int LineNum,
  int ColNum)
{
  StatusBar1->Panels->Items[0]->Text =
    String((int)Viewer1->CaretEffectivePos.Line + 1) + ":" +
    String((int)Viewer1->CaretEffectivePos.Col + 1);
  StatusBar1->Panels->Items[1]->Text =
    "Total: " + String((int)Viewer1->LineCount);
  StatusBar1->Panels->Items[2]->Text =
    "Top: " + String((int)Viewer1->TopLine + 1);
  StatusBar1->Panels->Items[3]->Text =
    "Bytes: " + String((int)Viewer1->FileSize);
  StatusBar1->Update();
  Clipboard1Click(this);
}
//---------------------------------------------------------------------------
void __fastcall TViewForm::About1Click(TObject *Sender)
{
  ViewAbout->ShowModal();
}
//---------------------------------------------------------------------------
void __fastcall TViewForm::FormCreate(TObject *Sender)
{
  TabExpansion1->Checked = Viewer1->ExpandTabs;
  DragAcceptFiles(Application->Handle, true);
  DragAcceptFiles(Handle, true);
  Application->OnMessage = AppMessage;
}
//---------------------------------------------------------------------------
void __fastcall TViewForm::TabExpansion1Click(TObject *Sender)
{
  TabExpansion1->Checked = !TabExpansion1->Checked;
  Viewer1->ExpandTabs = TabExpansion1->Checked;
}
//---------------------------------------------------------------------------
void __fastcall TViewForm::TabSize1Click(TObject *Sender)
{
  String S = String(Viewer1->TabSize);
  if (InputQuery("Tab Size", "Enter Tab Size", S))
    try {
      Viewer1->TabSize = (Word)S.ToInt();
    }
    catch (...) {
      return;
    }
}
//---------------------------------------------------------------------------
void __fastcall TViewForm::Find1Click(TObject *Sender)
{
  FindDialog1->Execute();
}
//---------------------------------------------------------------------------
void __fastcall TViewForm::FindDialog1Find(TObject *Sender)
{
  OptionSet.Clear();
  if (!FindDialog1->Options.Contains(frDown))
    OptionSet << soBackward;
  if (FindDialog1->Options.Contains(frMatchCase))
    OptionSet << soMatchCase;
  if (FindDialog1->Options.Contains(frWholeWord))
    OptionSet = OptionSet << soWholeWord;
  if (!Viewer1->Search(FindDialog1->FindText, OptionSet))
    MessageDlg("Search string '" + FindDialog1->FindText +
      "' not found", mtInformation, TMsgDlgButtons() << mbOK, 0);
  Search1Click(this);
}
//---------------------------------------------------------------------------
void __fastcall TViewForm::FindNext1Click(TObject *Sender)
{
  if (!Viewer1->Search(FindDialog1->FindText, OptionSet))
    MessageDlg("Search string '" + FindDialog1->FindText +
      "' not found", mtInformation, TMsgDlgButtons() << mbOK, 0);
  Search1Click(this);
}
//---------------------------------------------------------------------------
void __fastcall TViewForm::WordStarCommands1Click(TObject *Sender)
{
  WordStarCommands1->Checked = ! WordStarCommands1->Checked;
  DefaultController->EntryCommands->Table[0]->IsActive = WordStarCommands1->Checked;
}
//---------------------------------------------------------------------------
void __fastcall TViewForm::FormDestroy(TObject *Sender)
{
  DragAcceptFiles(Application->Handle, False);
  DragAcceptFiles(Handle, False);
}
//---------------------------------------------------------------------------
void __fastcall TViewForm::Print1Click(TObject *Sender)
{
  int I;
  String S;
  int X, Y, YD;
  int StartLine = 0, StopLine = 0;
  int StartCol, StopCol;

  if (Viewer1->GetSelection(StartLine, StartCol, StopLine, StopCol)) {
    PrintDialog1->Options << poSelection;
    PrintDialog1->PrintRange = prSelection;
  }
  else
    PrintDialog1->Options >> poSelection;
  if (!PrintDialog1->Execute()) return;

  // assign editor font to printer
  Printer()->Canvas->Font = Viewer1->FixedFont->Font;

  // get vertical line delta
  YD = Printer()->Canvas->TextHeight("Wg");

  // access printer handle to force selection of the font
  // into the printer DC-> Delphi doesn't do this automatically
  if (Printer()->Handle > 0) ;

  Screen->Cursor = crHourGlass;
  try {
    Y = 50;
    X = 50;
    if (StartLine == 0) StartLine = 1;
    if (StopLine == 0) StopLine = Viewer1->LineCount;
    Printer()->BeginDoc();
    for (I=StartLine;I<StopLine + 1;I++) {
      S = Viewer1->Lines[I];
      Printer()->Canvas->TextOut(X, Y, S);
      Y += YD;
      if (Y > Printer()->PageHeight - 50) {
        Y = 50;
        Printer()->NewPage();
      }
    }
    Printer()->EndDoc();
    Screen->Cursor = crDefault;
  }
  catch (...) {
    Screen->Cursor = crDefault;
  }
}
//---------------------------------------------------------------------------
